modules/http1/http1_filters.c

```diff
@@ -0,0 +1,373 @@
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+/*
+ * http_filter.c --- HTTP routines which either filters or deal with filters.
+ */
+
+#include "apr.h"
+#include "apr_strings.h"
+#include "apr_buckets.h"
+#include "apr_lib.h"
+#include "apr_signal.h"
+
+#define APR_WANT_STDIO          /* for sscanf */
+#define APR_WANT_STRFUNC
+#define APR_WANT_MEMFUNC
+#include "apr_want.h"
+
+#include "util_filter.h"
+#include "ap_config.h"
+#include "httpd.h"
+#include "http_config.h"
+#include "http_core.h"
+#include "http_protocol.h"
+#include "http_log.h"           /* For errors detected in basic auth common
+                                 * support code... */
+#include "apr_date.h"           /* For apr_date_parse_http and APR_DATE_BAD */
+#include "util_charset.h"
+#include "util_ebcdic.h"
+#include "util_time.h"
+
+#include "mod_core.h"
+#include "mod_http1.h"
+
+
+APLOG_USE_MODULE(http1);
+
+
+
+typedef struct response_filter_ctx {
+    int final_response_sent;    /* strict: a response status >= 200 was sent */
+    int discard_body;           /* the response is header only, discard body */
+    apr_bucket_brigade *tmpbb;
+} response_filter_ctx;
+
+AP_CORE_DECLARE_NONSTD(apr_status_t) ap_http1_response_out_filter(ap_filter_t *f,
+                                                                  apr_bucket_brigade *b)
+{
+    request_rec *r = f->r;
+    conn_rec *c = r->connection;
+    apr_bucket *e, *next = NULL;
+    response_filter_ctx *ctx = f->ctx;
+    apr_status_t rv = APR_SUCCESS;
+    core_server_config *conf = ap_get_core_module_config(r->server->module_config);
+    int strict = (conf->http_conformance != AP_HTTP_CONFORMANCE_UNSAFE);
+
+    AP_DEBUG_ASSERT(!r->main);
+
+    if (!ctx) {
+        ctx = f->ctx = apr_pcalloc(r->pool, sizeof(*ctx));
+    }
+
+    for (e = APR_BRIGADE_FIRST(b);
+         e != APR_BRIGADE_SENTINEL(b);
+         e = next)
+    {
+        next = APR_BUCKET_NEXT(e);
+
+        if (APR_BUCKET_IS_METADATA(e)) {
+
+            if (APR_BUCKET_IS_EOS(e)) {
+                if (!ctx->final_response_sent) {
+                    /* should not happen. do we generate a 500 here? */
+                }
+                ap_remove_output_filter(f);
+                goto pass;
+            }
+            else if (AP_BUCKET_IS_RESPONSE(e)) {
+                ap_bucket_response *resp = e->data;
+
+                ap_log_rerror(APLOG_MARK, APLOG_TRACE2, 0, r,
+                              "ap_http1_transcode_out_filter seeing response bucket status=%d",
```

Comment:
ap_http1_response_out_filter (found elsewhere in the PR too).

Reply:
Changed.